### ISSUE TITLE
Tcl / Tk failures for Python 3 Linux 64-bit wheel builds

### ISSUE THREAD

**Issue by matthew-brett** [matplotlib/matplotlib#8679]
For the last several days, for the wheel builds, I am seeing these failures from the Python 3 Linux 64-bit builds:

```
_____________________________ test_backend[tkagg] ______________________________
[gw0] linux -- Python 3.5.2 /venv/bin/python3.5
backend = 'tkagg'
    @pytest.mark.skipif("DISPLAY" not in os.environ,
                        reason="The DISPLAY environment variable is not set.")
    @pytest.mark.parametrize("backend", _get_available_backends())
    def test_backend(backend):
        environ = os.environ.copy()
        environ["MPLBACKEND"] = backend
        proc = Popen([sys.executable, "-c", _test_script], env=environ)
        # Empirically, 1s is not enough on Travis.
>       assert proc.wait(timeout=5) == 0
E       assert 1 == 0
E        +  where 1 = <bound method Popen.wait of <subprocess.Popen object at 0x7fda8c1f12e8>>(timeout=5)
E        +    where <bound method Popen.wait of <subprocess.Popen object at 0x7fda8c1f12e8>> = <subprocess.Popen object at 0x7fda8c1f12e8>.wait
/venv/lib/python3.5/site-packages/matplotlib/tests/test_backends_interactive.py:53: AssertionError
----------------------------- Captured stderr call -----------------------------
Traceback (most recent call last):
  File "<string>", line 4, in <module>
  File "/venv/lib/python3.5/site-packages/matplotlib/pyplot.py", line 544, in figure
    **kwargs)
  File "/venv/lib/python3.5/site-packages/matplotlib/backends/backend_tkagg.py", line 80, in new_figure_manager
    return new_figure_manager_given_figure(num, figure)
  File "/venv/lib/python3.5/site-packages/matplotlib/backends/backend_tkagg.py", line 88, in new_figure_manager_given_figure
    window = Tk.Tk(className="matplotlib")
  File "/usr/lib/python3.5/tkinter/__init__.py", line 1871, in __init__
    self.tk = _tkinter.create(screenName, baseName, className, interactive, wantobjects, useTk, sync, use)
_tkinter.TclError: no display name and no $DISPLAY environment variable
!!!!!!!!!!!! xdist.dsession.Interrupted: stopping after 1 failures !!!!!!!!!!!!!
======== 1 failed, 398 passed, 156 skipped, 5 xfailed in 141.40 seconds ========
```

See: 

* example failure log: https://travis-ci.org/MacPython/matplotlib-wheels/jobs/237087622
* build matrix: https://travis-ci.org/MacPython/matplotlib-wheels/builds/237087618

Just superficially, it seems that the test decorator should be skipping this test.  Any ideas what's going on here?

**Comment by dstansby** [comment]
The test was added by #8660

**Comment by anntzer** [comment]
Can you try whether replacing `"DISPLAY" not in os.environ` by `not os.environ.get("DISPLAY")` (empty values for environment variables...) would help?

**Comment by matthew-brett** [comment]
Sadly no, it didn't help:  https://travis-ci.org/MacPython/matplotlib-wheels/jobs/237287904#L646

I also had a look at the image that it's using for testing - https://hub.docker.com/r/matthewbrett/trusty - `DISPLAY` is not set.

**Comment by matthew-brett** [comment]
Debug print of `python -c "import os; print(not 'DISPLAY' in os.environ)"` before tests are run gives `True`.

https://travis-ci.org/MacPython/matplotlib-wheels/jobs/237292458#L579

**Comment by anntzer** [comment]
What happens with
```
diff --git a/lib/matplotlib/tests/test_backends_interactive.py b/lib/matplotlib/tests/test_backends_interactive.py
index 1b763ea06..f38ff0a6f 100644
--- a/lib/matplotlib/tests/test_backends_interactive.py
+++ b/lib/matplotlib/tests/test_backends_interactive.py
@@ -41,7 +41,7 @@ plt.show()
 """
 
 
-@pytest.mark.skipif("DISPLAY" not in os.environ,
+@pytest.mark.skipif(print("===", repr(os.environ.get("DISPLAY")), "===") or not os.environ.get("DISPLAY"),
                     reason="The DISPLAY environment variable is not set.")
 @pytest.mark.parametrize("backend", _get_available_backends())
 def test_backend(backend):
```

trying to see if DISPLAY somehow becomes set during test collection...

**Comment by matthew-brett** [comment]
Actually, this assert passes inside the test function:

```
assert environ.get('DISPLAY') is None
```

https://travis-ci.org/MacPython/matplotlib-wheels/builds/237425182#L657

I wonder whether the skip decorator is failing to work for all the parametrized backend tests?  I mean, maybe the double decoration is the problem?

```
@pytest.mark.skipif("DISPLAY" not in os.environ,
                    reason="The DISPLAY environment variable is not set.")
@pytest.mark.parametrize("backend", _get_available_backends())
def test_backend(backend):
```

I tried reversing the decorators without change.

**Comment by anntzer** [comment]
Well the tests to pass on OSX on Travis for us and they do not have an X server (see before_script in .travis.yml) so the skipping must work fine...  Still, try something like
```
diff --git a/lib/matplotlib/tests/test_backends_interactive.py b/lib/matplotlib/tests/test_backends_interactive.py
index 1b763ea06..16d276876 100644
--- a/lib/matplotlib/tests/test_backends_interactive.py
+++ b/lib/matplotlib/tests/test_backends_interactive.py
@@ -21,10 +21,12 @@ import pytest
 
 def _get_available_backends():
     return [
-        pytest.mark.skipif(sys.version_info < (3,)
-                           or importlib.util.find_spec(module_name) is None,
-                           reason="Could not import {!r}".format(module_name))(
-                               backend)
+        pytest.mark.skipif(
+            not os.environ.get("DISPLAY")
+            or sys.version_info < (3,)
+            or importlib.util.find_spec(module_name) is None,
+            reason="No $DISPLAY, or could not import {!r}".format(module_name))(
+                backend)
         for module_name, backend in [
                 ("PyQt5", "qt5agg"),
                 ("tkinter", "tkagg"),
@@ -41,8 +43,6 @@ plt.show()
 """
 
 
-@pytest.mark.skipif("DISPLAY" not in os.environ,
-                    reason="The DISPLAY environment variable is not set.")
 @pytest.mark.parametrize("backend", _get_available_backends())
 def test_backend(backend):
     environ = os.environ.copy()
```

**Comment by matthew-brett** [comment]
Yes, that last patch works : https://travis-ci.org/MacPython/matplotlib-wheels/builds/237893723

I guess you could rename the function to `_get_testable_interactive_backends`.

**Comment by anntzer** [comment]
pr opened

**Comment by anntzer** [comment]
I assumed this is closed by #8691, please request a reopen otherwise.
